internal/functions/stencil.go

```diff
@@ -0,0 +1,178 @@
+func NewStencil(m *configuration.ServiceManifest, mods []*modules.Module) *Stencil {
+	return &Stencil{m, mods, make([]*Template, 0), nil}
+}
+
+// Stencil provides the basic functions for
+// stencil templates
+type Stencil struct {
+	m *configuration.ServiceManifest
+
+	// Modules is a list of modules used in this stencil render
+	Modules []*modules.Module
+
+	// Templates is a list of all templates that this renderer rendered.
+	Templates []*Template
+
+	// Template is the current template that is being rendered by this
+	// renderer.
+	Template *Template
+}
+
+// GenerateLockfile generates a stencil.Lockfile based
+// on the current state of the renderer.
+func (s *Stencil) GenerateLockfile() *stencil.Lockfile {
+	l := &stencil.Lockfile{
+		Version:   app.Info().Version,
+		Generated: time.Now().UTC(),
+	}
+
+	for _, tpl := range s.Templates {
+		for _, f := range tpl.Files {
+			l.Files = append(l.Files, &stencil.LockfileFileEntry{
+				Name:     f.Name(),
+				Template: tpl.Path,
+				Module:   tpl.Module.Name,
+			})
+		}
+	}
+
+	for _, m := range s.Modules {
+		l.Modules = append(l.Modules, &stencil.LockfileModuleEntry{
+			Name:    m.Name,
+			URL:     m.URI,
+			Version: m.Version,
+		})
+	}
+
+	return l
+}
+
+// Render renders all templates using the ServiceManifest that was
+// provided to stencil at creation time. The rendered templates are
+// available at Templates on the Stencil struct.
+// TODO(jaredallard): Do we want to just return []*Template?
+func (s *Stencil) Render(ctx context.Context) error {
+	args, err := s.makeTemplateParameters()
+	if err != nil {
+		return err
+	}
+
+	tpls, err := s.getTemplates(ctx)
+	if err != nil {
+		return err
+	}
+	// Add the templates to their modules template to allow them to be able to access
+	// functions declared in the same module
+	for _, t := range tpls {
+		if err := t.Parse(s); err != nil {
+			return errors.Wrapf(err, "failed to parse template %q", t.ImportPath())
+		}
+	}
+
+	// Now we render each file
+	for _, t := range tpls {
+		// change the active template
+		s.Template = t
+
+		if err := t.Render(s, args); err != nil {
+			return errors.Wrapf(err, "failed to render template %q", t.ImportPath())
+		}
+
+		// append the rendered template to our list of templates processed
+		s.Templates = append(s.Templates, t)
+	}
+
+	return nil
+}
+
+// makeTemplateParameters creates the map to be provided to the templates.
+func (s *Stencil) makeTemplateParameters() (map[string]interface{}, error) {
+	// TODO(jaredallard): head branch
+	boxConf, err := box.LoadBox()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to load box config")
+	}
+
+	return map[string]interface{}{
+		"Metadata": map[string]string{
+			"Generator": app.Info().Name,
+			"Version":   app.Info().Version,
+		},
+		"App": map[string]string{
+			"Name": s.m.Name,
+		},
+		"Repository": map[string]string{
+			"HeadBranch": "main",
+		},
+		"Box": boxConf,
```

Comment:
what is the behavior if i, as a module, decide one of my argument keys should be named `App`? does this default get overwritten?

Reply:
I think arguments can only be accessed via `stencil.Args` ?  Hopefully we are not injecting arguments directly.

It would be nice if all of these were strongly typed (i.e. no `map[string]string{...}`)...

Also, I think `Metadata/App/Repository` don't seem like a clear distinction.. can you summarize what goes where?

Also, I think `metadata/repo/app` seems less descriptive... 

I tend to think in terms of things that are properties of the installation (on disk and not related directly to service.yaml/stencil), things that are a property of the runtime (i.e. stencil host environment + invocation, possibly golang version if we ever need it but also stuff like standard copyright notices) and things that are stencil config (but not specific to a module, things like app name that go into service.yaml):

* install (anything about the current installation: repo name, branch, tag etc).
* runtime (anything about the current runtime environment: calling app name, calling app version, whether running in CI or local etc; access to runtime flags etc could happen via this as well such as debug logging)
* config (any global shared service.yaml config such as app name)

Alternative break downs are ok -- but would prefer if we clearly pinned down the boundaries up-front.

Reply:
no, these are different. Arguments are accessed via `stencil.Arg(name)` or `stencil.Args().path`, while these come from `.App` or `.Metadata`, so on.

Reply:
Makes sense, thats my own misunderstanding then

Reply:
Ah -- didn't see your comment @vk-outreach (thanks Github). Totally agree, already making them strongly typed and with your distinctions. A much better way of doing what I wanted to do :D

Reply:
Changed this up a little bit... PTAL! (`internal/codegen/values.go`)